### ISSUE TITLE
shutdown crash in smem even when not using smem

### ISSUE THREAD

**Issue by marinier** [SoarGroup/Soar#243]
I’m running with Soar built from the development branch (commit 6d38396, Windows 7 64 with VS 2013 Express). With the simple application below, I get a crash during shutdown in sml_ClientKernel.cpp line 3359 (in smem_close()) – it appears to be trying to disconnect from the smem db, which is strange since I never turned smem on.

    int main(int argc, char **argv) {
           Kernel* const kern = Kernel::CreateKernelInNewThread();
           Agent* const agent = kern->CreateAgent("demo");
           kern->RunAllAgents(1);
           kern->Shutdown();
           return 0;
    }

From Mazin: "Steven and I did notice yesterday that Soar attempts to attach to smem every decision cycle regardless of whether smem is on, which seemed odd and could explain why smem might be involved.  Anyway, this seems like a clear bug."

**Comment by mazina** [comment]
Hmm...I build that test, but it does not crash for me.  It ran one step and exited.
```
~/Soar/SoarSandbox> ./SML_Test

     0: ==>S: S1
     1:    ==>S: S5 (state no-change)
~/Soar/SoarSandbox>
```

Any ideas?

**Comment by AlexNickels** [comment]
Just as an odd note: I observed shutdown bugs on both memories while working on the android port.  In my case it tries to disconnect multiple times.

**Comment by marinier** [comment]
It could be that we have different setups and that's making some kind of a
difference. I was using Visual Studio 2013 Express on Windows 7 64-bit.
Were you using a different configuration? Also, do you know if it is
getting to the line I indicated and just not crashing, vs. not even getting
to that line at all?

**Comment by mazina** [comment]
I built mine on OSX.  

I can check if it calls that function.  I'd guess it does since smem seems to have been coded to always try to attach to the database even when it's not on.

**Comment by mazina** [comment]
Yes, it is calling smem_close() on exit.

**Comment by marinier** [comment]
Does it hit line 3359 in smem_close()? The first thing smem_close does is check if it needs to do anything. For me, it decides it does and that's when it gets into trouble. I believe that calling this every time during shutdown is probably correct, but it's the check it's doing that is somehow messed up (either the check is wrong, or something that is being checked isn't set right elsewhere).

It could also be that the implementation of sqlite on OS X gracefully handles this situation (or gets lucky) while the Windows implementation does not.

**Comment by mazina** [comment]
It does hit that line.  I even stepped through it down to the sql lite
code, and there seemed to be a valid database there that it had no issue
properly closing. Maybe the database object is somehow getting corrupted on
Windows?

**Comment by mazina** [comment]
OK, I figured out the bug.  It's actually the same thing that has been causing the random unit test crash on Windows that Nathan noticed and was trying to debug.  

I overcame my fears of running VisualStudio in a Windows VM, and it seems that memory was being corrupted when it deleted the variable thisAgent->lastCue on the line before that.  lastCue contained garbage and, for some reason on Windows, deleting that bad address caused the my_db variable to consistently become corrupted.  

The lastCue variable seems to be something that is only used in one of @bluechill 's new unit tests that he ported from JSoar.  The problem came out of something weird he was doing initializing that variable only when queries were possible.  In your example and the crashing unit tests, I think that init code never gets executed because smem is off.  I bet on Mac and Unix, the compiler automatically initialize last_cue to NULL, so this bug doesn't become apparent.  In any case, I added some code to initialize it to NULL on agent creation as well, and the problem went away.

**Comment by marinier** [comment]
Awesome! Thanks for figuring this out. I believe the behavior of
uninitialized variables is undefined. Setting them to null is sort of
helpful, but can end up hiding real bugs like this, so I actually prefer
VS's behavior here :)

**Comment by garfieldnate** [comment]
Seems like better behavior would be to detect the possibility of an uninitialized variable and warn about it. That's what Java does, anyway. Is there no flag to make this warn for Clang or gcc?

**Comment by marinier** [comment]
Yes, all the compilers can report that warning, although I'm not sure it's
on by default in gcc. It is on by default in Visual Studio. The problem is
that all the compilers report so many spurious warnings that any useful
ones get lost in the shuffle. At one point in the past we actually got rid
of all or almost all of the spurious warnings across all the compilers we
were using (and discovered and fixed a couple real issues along the way),
but it's actually a lot of work to do that, and clearly hasn't been
maintained. Maybe it would be useful to do that again, at least for the 1
or 2 compilers that are most commonly used for development these days.

**Comment by garfieldnate** [comment]
I think it would be worth it to get rid of all spurious warnings, though I know it takes time. I'm always worried that I broke something when I see them.

**Comment by marinier** [comment]
I'm actually surprised the jenkins builds didn't catch this. I would have
thought this would have made many of the tests on windows crash. If it
didn't, maybe this simple test case should be added. Or maybe jenkins is
only doing builds and not tests?

**Comment by mazina** [comment]
There's a story behind it.  For a good while, Jenkins was down.  To be honest, no one was doing any kernel work, so we didn't even notice they were down for a long time.  Alex got them back up during the summer and added some new things, among which was his new testing framework.  Unfortunately, some of the tests he ported from jSoar still fail (some for known, valid reasons), and we haven't had the time or resources to fix them yet.  I thought he had changed the jenkins script/testing framework so that it just ignored those failing tests, but it looks like it was ignoring all the unit tests.  

I've changed it to use the old unit tests again for now.

**Comment by bluechill** [comment]
I ran it with clang's address sanitizer which should have caught that, apparently it didn't.  And Jenkins should be using the new tests because they are *far* more extensive.  And just disable the unit tests within the testing framework not the whole unit tests...  It's really simple... just comment out the line in the unit testing class and make a note that it's still failing.  I don't have time to do much of anything right now but just switching to the other unit tests is a really bad idea.

**Comment by mazina** [comment]
There's a story behind it.   During the summer, Alex added some new things
to the Jenkins script, among which was his new testing framework.
Unfortunately, some of the tests he ported from jSoar still fail (some for
known, valid reasons), and we haven't had the time or resources to fix them
yet.  I thought we had changed the jenkins script/testing framework so that
it just ignored those failing tests, but it looks like it was ignoring all
the unit tests.

I could be slightly off about that.   I'm a little hazy on the details now,
but they might be disabled for another reason.  Jenkins went down recently
when there was a power outage in the server room.  When we got them back
up, I had several problems with the script. It's possible the new testing
framework was disabled because of a general problem getting jenkins to run
it at all, not because those few tests were failing.  I also had issues
getting it to run the performance tests, so that's also currently disabled.

Anyway, I've changed it to use the old unit tests again for now.  Once I
have time to mess around with the Alex's new tests, I'll re-enable them.
